src/macaron/repo_finder/commit_finder.py

```diff
@@ -0,0 +1,160 @@
+# Copyright (c) 2023 - 2023, Oracle and/or its affiliates. All rights reserved.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.
+
+"""This module contains the logic for matching PackageURL versions to repository commits via the tags they contain."""
+import logging
+import re
+from re import Pattern
+
+from git import TagReference
+from packageurl import PackageURL
+from pydriller import Git
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+ALPHANUMERIC = "0-9a-z"
+PREFIX = "(?:.*)"
+INFIX = "[^0-9]*"
+split_pattern = re.compile(f"[^{ALPHANUMERIC}]", flags=re.IGNORECASE)
+validation_pattern = re.compile(f"[{ALPHANUMERIC}]+", flags=re.IGNORECASE)
+
+
+def get_commit_from_version(git_obj: Git, purl: PackageURL) -> tuple[str, str]:
+    """Try to find the matching commit in a repository of a given version via tags.
+
+    The version of the passed PackageURL is used to match with the tags in the target repository.
+
+    Parameters
+    ----------
+    git_obj: Git
+        The repository.
+    purl: PackageURL | None
+        The PURL of the artifact.
+
+    Returns
+    -------
+    tuple[str, str]
+        The branch name and digest as a tuple.
+    """
+    if purl.version is None:
+        logger.debug("Missing version for artifact: %s", purl.name)
+        return "", ""
+    logger.debug("Searching for commit of artifact version using tags: %s@%s", purl.name, purl.version)
+    # Create version based regex from the PackageURL for matching against tags.
+    # The version is split on non-alphanumeric characters to separate the version parts from the non-version parts.
+    # e.g. 1.2.3-DEV -> [1, 2, 3, DEV]
+    split = split_pattern.split(purl.version)
+    logger.debug("Split version: %s", split)
+    if not split:
+        split = [purl.version]
+    this_version_pattern = ""
+    for part in split:
+        # Validate the split part by checking it is only comprised of alphanumeric characters.
+        valid = validation_pattern.match(part)
```

Comment:
There was a slight error in the regex that I've now fixed. It is supposed to split on any number of separators, so your example above would produce `['1', '2', '3', 'A']`.
This extra validation was added more for protection against shenanigans, as the version string is essentially user input, and parts of it will be directly used in a new regex pattern. I'm not sure whether that kind of injection matters, or if it's not already too late at that point, but I thought making an attempt was still worthwhile.